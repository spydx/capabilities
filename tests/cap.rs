use async_trait::async_trait;

#[async_trait]
pub trait Capability<Operation> {
    type Data;
    type Error;

    async fn perform(&self, _: Operation) -> Result<Self::Data, Self::Error>;
}

macro_rules! cap {
    ($name:ident for $type:ty, composing $({$operation:ty, $d:ty, $e:ty}),+) => {
        #[async_trait]
        pub trait $name: $(Capability<$operation, Data = $d, Error = $e>+)+ {}

        #[async_trait]
        impl $name for $type {}
    };
}
struct Orders {
    id: i32, 
    name: String,
}

struct Database;
struct DatabaseError;

struct Read<T>{data: T}

cap!(CanReadOrders for Database, composing { Read<String>, Orders, DatabaseError} );


#[tokio::main]
async fn main() -> Result<(), std::io::Error> {
    
    Ok(())
}
