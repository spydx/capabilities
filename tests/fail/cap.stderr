error[E0277]: the trait bound `Database: Capability<Read<String>>` is not satisfied
  --> tests/fail/cap.rs:32:6
   |
32 | cap!(CanReadOrders for Database, composing { Read<String>, Orders, DatabaseError} );
   |      ^^^^^^^^^^^^^ the trait `Capability<Read<String>>` is not implemented for `Database`
   |
note: required by a bound in `CanReadOrders`
  --> tests/fail/cap.rs:14:28
   |
14 |         pub trait $name: $(Capability<$operation, Data = $d, Error = $e>+)+ {}
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `CanReadOrders`
...
32 | cap!(CanReadOrders for Database, composing { Read<String>, Orders, DatabaseError} );
   | -----------------------------------------------------------------------------------
   | |    |
   | |    required by a bound in this
   | in this macro invocation
   = note: this error originates in the macro `cap` (in Nightly builds, run with -Z macro-backtrace for more info)
